### DISCUSSION TITLE
Displaying Modflow console output in VS code

### DISCUSSION THREAD

**Discussion by BojanMilinic** [gmdsi/GMDSI_notebooks#134] [General]
Hi all. 

Is there a way to look at the modflow console output when running a jupyter notebook in VS code? 

I noticed if you just use Anaconda Prompt and open a Jupyter notebook, running a modflow simulation using pyemu results in the Anaconda prompt terminal displaying the Modflow console output. For example, "Solving:  Stress period:     1    Time step:     1"

But when running a jupyter notebook in vs code, running a modflow simulation displays no modflow output in the terminal.  

I also tried going to the "OUTPUT" tab in vs code but still nothing pops up when running a simulation. 

Either way, thanks for teaching this course,
Bojan

**Comment by mnfienen** [comment]
Hey @BojanMilinic  - I think VSCode may default to suppressing console output. To be safe, if you want to see the console, you might need to run through the terminal.

**Comment by p-ortega** [comment]
@BojanMilinic  I also use vscode and it seems `os.system()` doesn't transfer the stdout to vscode's interactive window. 

Another workaround is the same func but using subprocess ..

```
import platform
import subprocess


def run2(cmd_str, cwd=".", verbose=False):
    """an OS agnostic function to execute a command line

    Args:
        cmd_str (`str`): the str to execute with `os.system()`

        cwd (`str`, optional): the directory to execute the command in.
            Default is ".".
        verbose (`bool`, optional): flag to echo to stdout the  `cmd_str`.
            Default is `False`.

    Notes:
        uses `platform` to detect OS and adds .exe suffix or ./ prefix as appropriate
        if `os.system` returns non-zero, an exception is raised

    Example::

        pyemu.os_utils.run("pestpp-ies my.pst",cwd="template")

    """
    bwd = os.getcwd()
    os.chdir(cwd)
    try:
        exe_name = cmd_str.split()[0]
        if "window" in platform.platform().lower():
            if not exe_name.lower().endswith("exe"):
                raw = cmd_str.split()
                raw[0] = exe_name + ".exe"
                cmd_str = " ".join(raw)
        else:
            if exe_name.lower().endswith("exe"):
                raw = cmd_str.split()
                exe_name = exe_name.replace(".exe", "")
                raw[0] = exe_name
                cmd_str = "{0} {1} ".format(*raw)
            if os.path.exists(exe_name) and not exe_name.startswith("./"):
                cmd_str = "./" + cmd_str

    except Exception as e:
        os.chdir(bwd)
        raise Exception("run() error preprocessing command line :{0}".format(str(e)))
    if verbose:
        print("run():{0}".format(cmd_str))

    try:
        ret_val = subprocess.run([i for i in cmd_str.split(' ')], capture_output=True)
        print(ret_val.stdout.decode('utf-8'))
    except Exception as e:
        os.chdir(bwd)
        raise Exception("run() raised :{0}".format(str(e)))
    os.chdir(bwd)

    if "window" in platform.platform().lower():
        if ret_val.returncode != 0:
            raise Exception("run() returned non-zero: {0}".format(ret_val.returncode))
    else:
        estat = os.WEXITSTATUS(ret_val.returncode)
        if estat != 0 or ret_val != 0:
            raise Exception("run() returned non-zero: {0},{1}".format(estat,ret_val.returncode))
```

**Reply by briochh** [reply]
hey @p-ortega, This is a good call. I know a few people who have been left a bit "blind" when launching from within notebooks.  If you think subprocess is a sensible direction to go in for spawning out system commands we would be happy to receive a pull request over on the [pyemu repo](https://github.com/pypest/pyemu). There may be other instances where subprocess variants can be subbed in. Keen to hear your thoughts on the pros and cons of the different options.

**Comment by p-ortega** [comment]
Hi @briochh,

I don't know all the implications of moving from `os.system` to `subprocess`, but it seems to be the preferred choice from the dodgy StackOverflow threads I frequent. Less overhead in Windows, avoiding running a shell instance (security vulnerability?), and more control over the processes seem to be the main reasons.

Now, for me personally, one interesting option with `subprocess.Popen` would be to run pestpp detached from the terminal. This would (hopefully) prevent accidentally killing someone's else run by closing the terminal (and avoid releasing the dogs of hell).

Lemme know what you think ..

**Comment by p-ortega** [comment]
I think this works now
